File: deepeval/metrics/conversational_g_eval/conversational_g_eval.py
"""A slightly modified tailored version of the LLM evaluated metric based on the GEval framework: https://arxiv.org/pdf/2303.16634.pdf"""

from openai.types.chat.chat_completion import ChatCompletion
from typing import Optional, List, Tuple, Union, Dict, Type
import math
from deepeval.metrics import BaseConversationalMetric
from deepeval.metrics.g_eval.utils import (
    Rubric,
    construct_conversational_g_eval_turn_params_string,
    construct_non_turns_test_case_string,
    format_rubrics,
)
from deepeval.test_case import (
    TurnParams,
    ConversationalTestCase,
)
from deepeval.metrics.conversational_g_eval.template import (
    ConversationalGEvalTemplate,
)
from deepeval.utils import get_or_create_event_loop, prettify_list
from deepeval.metrics.utils import (
    check_conversational_test_case_params,
    construct_verbose_logs,
    a_gen_and_extract,
    gen_and_extract,
    trimAndLoadJson,
    initialize_model,
    convert_turn_to_dict,
)
from deepeval.models import DeepEvalBaseLLM
from deepeval.metrics.indicator import metric_progress_indicator
import deepeval.metrics.conversational_g_eval.schema as cgschema
from deepeval.metrics.api import metric_data_manager


class ConversationalGEval(BaseConversationalMetric):
    def __init__(
        self,
        name: str,
        evaluation_params: Optional[List[TurnParams]] = None,
        criteria: Optional[str] = None,
        evaluation_steps: Optional[List[str]] = None,
        model: Optional[Union[str, DeepEvalBaseLLM]] = None,
        threshold: float = 0.5,
        rubric: Optional[List[Rubric]] = None,
        async_mode: bool = True,
        strict_mode: bool = False,
        verbose_mode: bool = False,
        evaluation_template: Type[
            ConversationalGEvalTemplate
        ] = ConversationalGEvalTemplate,
        _include_g_eval_suffix: bool = True,
    ):
        if evaluation_params is not None and len(evaluation_params) == 0:
            raise ValueError("evaluation_params cannot be an empty list.")

        self.name = name
        if evaluation_params is None:
            evaluation_params = [TurnParams.CONTENT, TurnParams.ROLE]

        if TurnParams.CONTENT not in evaluation_params:
            evaluation_params.append(TurnParams.CONTENT)
        if TurnParams.ROLE not in evaluation_params:
            evaluation_params.append(TurnParams.ROLE)

        self.evaluation_params = evaluation_params

        # Check if both criteria and evaluation_steps are not None at the same time
        if criteria is None and evaluation_steps is None:
            raise ValueError(
                "Either 'criteria' or 'evaluation_steps' must be provided."
            )

        # Check if criteria is provided, it cannot be an empty string
        if criteria is not None and not criteria.strip():
            raise ValueError("Criteria provided cannot be an empty string.")

        # Check if evaluation_steps is provided, it cannot be an empty list
        if evaluation_steps is not None and len(evaluation_steps) == 0:
            raise ValueError(
                "'evaluation_steps' must not be an empty list. Either omit evaluation steps or include a non-empty list of steps."
            )

        self.criteria = criteria
        self.rubric = rubric
        self.model, self.using_native_model = initialize_model(model)
        self.evaluation_model = self.model.get_model_name()
        self.evaluation_steps = evaluation_steps
        self.threshold = 1 if strict_mode else threshold
        self.strict_mode = strict_mode
        self.async_mode = async_mode
        self.verbose_mode = verbose_mode
        self.evaluation_template = evaluation_template
        self._include_g_eval_suffix = _include_g_eval_suffix

    def measure(
        self,
        test_case: ConversationalTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
        _log_metric_to_confident: bool = True,
    ) -> float:
        check_conversational_test_case_params(
            test_case, self.evaluation_params, self
        )

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self, _show_indicator=_show_indicator, _in_component=_in_component
        ):
            if self.async_mode:
                loop = get_or_create_event_loop()
                loop.run_until_complete(
                    self.a_measure(
                        test_case,
                        _show_indicator=False,
                        _in_component=_in_component,
                        _log_metric_to_confident=_log_metric_to_confident,
                    )
                )
            else:
                self.evaluation_steps: List[str] = (
                    self._generate_evaluation_steps()
                )
                g_score, reason = self.evaluate(test_case)
                self.reason = reason
                self.score = float(g_score) / 10
                self.score = (
                    0
                    if self.strict_mode and self.score < self.threshold
                    else self.score
                )
                self.success = self.score >= self.threshold
                self.verbose_logs = construct_verbose_logs(
                    self,
                    steps=[
                        f"Criteria:\n{self.criteria}",
                        f"Evaluation Steps:\n{prettify_list(self.evaluation_steps)}",
                        f"Rubric:\n{format_rubrics(self.rubric)}",
                        f"Score: {self.score}\nReason: {self.reason}",
                    ],
                )
                if _log_metric_to_confident:
                    metric_data_manager.post_metric_if_enabled(
                        self, test_case=test_case
                    )

            return self.score

    async def a_measure(
        self,
        test_case: ConversationalTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
        _log_metric_to_confident: bool = True,
    ) -> float:
        check_conversational_test_case_params(
            test_case, self.evaluation_params, self
        )

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self,
            async_mode=True,
            _show_indicator=_show_indicator,
            _in_component=_in_component,
        ):
            self.evaluation_steps: List[str] = (
                await self._a_generate_evaluation_steps()
            )
            g_score, reason = await self._a_evaluate(test_case)
            self.reason = reason
            self.score = float(g_score) / 10
            self.score = (
                0
                if self.strict_mode and self.score < self.threshold
                else self.score
            )
            self.success = self.score >= self.threshold
            self.verbose_logs = construct_verbose_logs(
                self,
                steps=[
                    f"Criteria:\n{self.criteria}",
                    f"Evaluation Steps:\n{prettify_list(self.evaluation_steps)}",
                    f"Rubric:\n{format_rubrics(self.rubric)}",
                    f"Score: {self.score}\nReason: {self.reason}",
                ],
            )
            if _log_metric_to_confident:
                metric_data_manager.post_metric_if_enabled(
                    self, test_case=test_case
                )

            return self.score

    async def _a_generate_evaluation_steps(self) -> List[str]:
        if self.evaluation_steps:
            return self.evaluation_steps

        g_eval_params_str = construct_conversational_g_eval_turn_params_string(
            self.evaluation_params
        )
        prompt = self.evaluation_template.generate_evaluation_steps(
            criteria=self.criteria, parameters=g_eval_params_str
        )

        return await a_gen_and_extract(
            self,
            prompt,
            cgschema.Steps,
            extract_schema=lambda r: r.steps,
            extract_json=lambda d: d["steps"],
        )

    def _generate_evaluation_steps(self) -> List[str]:
        if self.evaluation_steps:
            return self.evaluation_steps

        g_eval_params_str = construct_conversational_g_eval_turn_params_string(
            self.evaluation_params
        )
        prompt = self.evaluation_template.generate_evaluation_steps(
            criteria=self.criteria, parameters=g_eval_params_str
        )

        return gen_and_extract(
            self,
            prompt,
            cgschema.Steps,
            extract_schema=lambda r: r.steps,
            extract_json=lambda d: d["steps"],
        )

    async def _a_evaluate(
        self, test_case: ConversationalTestCase
    ) -> Tuple[Union[int, float], str]:
        test_case_content = construct_non_turns_test_case_string(
            self.evaluation_params, test_case
        )
        g_eval_params_str = construct_conversational_g_eval_turn_params_string(
            self.evaluation_params
        )
        if not self.strict_mode:
            rubric_str = format_rubrics(self.rubric) if self.rubric else None
            prompt = self.evaluation_template.generate_evaluation_results(
                evaluation_steps=self.number_evaluation_steps(),
                test_case_content=test_case_content,
                turns=[
                    convert_turn_to_dict(turn, self.evaluation_params)
                    for turn in test_case.turns
                ],
                parameters=g_eval_params_str,
                rubric=rubric_str,
            )
        else:
            prompt = self.evaluation_template.generate_evaluation_results(
                evaluation_steps=self.number_evaluation_steps(),
                test_case_content=test_case_content,
                turns=[
                    convert_turn_to_dict(turn, self.evaluation_params)
                    for turn in test_case.turns
                ],
                parameters=g_eval_params_str,
            )
        try:
            result, cost = await self.model.a_generate_raw_response(
                prompt, top_logprobs=20
            )
            self.evaluation_cost += cost
            data = trimAndLoadJson(result.choices[0].message.content, self)

            reason = data["reason"]
            score = data["score"]
            if self.strict_mode:
                return score, reason

            try:
                weighted_summed_score = self.generate_weighted_summed_score(
                    score, result
                )
                return weighted_summed_score, reason
            except (KeyError, AttributeError, TypeError, ValueError):
                return score, reason
        except (
            AttributeError
        ):  # This catches the case where a_generate_raw_response doesn't exist.

            return await a_gen_and_extract(
                self,
                prompt,
                cgschema.ReasonScore,
                extract_schema=lambda r: (r.score, r.reason),
                extract_json=lambda d: (d["score"], d["reason"]),
            )

    def evaluate(
        self, test_case: ConversationalTestCase
    ) -> Tuple[Union[int, float], str]:
        test_case_content = construct_non_turns_test_case_string(
            self.evaluation_params, test_case
        )
        g_eval_params_str = construct_conversational_g_eval_turn_params_string(
            self.evaluation_params
        )
        if not self.strict_mode:
            rubric_str = format_rubrics(self.rubric) if self.rubric else None
            prompt = self.evaluation_template.generate_evaluation_results(
                evaluation_steps=self.number_evaluation_steps(),
                test_case_content=test_case_content,
                turns=[
                    convert_turn_to_dict(turn, self.evaluation_params)
                    for turn in test_case.turns
                ],
                parameters=g_eval_params_str,
                rubric=rubric_str,
            )
        else:
            prompt = self.evaluation_template.generate_evaluation_results(
                evaluation_steps=self.number_evaluation_steps(),
                test_case_content=test_case_content,
                turns=[
                    convert_turn_to_dict(turn, self.evaluation_params)
                    for turn in test_case.turns
                ],
                parameters=g_eval_params_str,
            )
        try:
            result, cost = self.model.generate_raw_response(
                prompt, top_logprobs=20
            )
            self.evaluation_cost += cost
            data = trimAndLoadJson(result.choices[0].message.content, self)

            reason = data["reason"]
            score = data["score"]
            if self.strict_mode:
                return score, reason

            try:
                weighted_summed_score = self.generate_weighted_summed_score(
                    score, result
                )
                return weighted_summed_score, reason
            except (KeyError, AttributeError, TypeError, ValueError):
                return score, reason
        except AttributeError:
            # This catches the case where a_generate_raw_response doesn't exist.
            return gen_and_extract(
                self,
                prompt,
                cgschema.ReasonScore,
                extract_schema=lambda r: (r.score, r.reason),
                extract_json=lambda d: (d["score"], d["reason"]),
            )

    def generate_weighted_summed_score(
        self, raw_score: int, raw_response: ChatCompletion
    ) -> Union[int, float]:
        generated_logprobs = raw_response.choices[0].logprobs.content
        # First, locate the token that we care for logprobs, i.e., the token matching the score
        score_logprobs = None
        for token_logprobs in generated_logprobs:
            if token_logprobs.token == str(raw_score):
                score_logprobs = token_logprobs
                break
        # Then, calculate the score based on the logprobs
        token_linear_probability: Dict[int, float] = {}
        sum_linear_probability = 0
        # Filter out tokens with <1% linear probability, i.e., logprobs < math.log(0.01)
        min_logprob = math.log(0.01)
        for token_logprob in score_logprobs.top_logprobs:
            logprob = token_logprob.logprob

            # Filter out low probability tokens
            if logprob < min_logprob:
                continue
            # Filter out non-decimal token to prevent errors in later int(token) conversion
            if not token_logprob.token.isdecimal():
                continue

            # Calculate the linear probability
            linear_prob = math.exp(logprob)
            token_score = int(token_logprob.token)
            if token_linear_probability.get(token_score):
                token_linear_probability[token_score] += linear_prob
            else:
                token_linear_probability[token_score] = linear_prob
            sum_linear_probability += linear_prob

        sum_of_weighted_scores = 0.0
        for score, prob in token_linear_probability.items():
            sum_of_weighted_scores += score * prob

        # Scale the sum of linear probability to 1
        weighted_summed_score = sum_of_weighted_scores / sum_linear_probability
        return weighted_summed_score

    def number_evaluation_steps(self):
        evaluation_steps = """"""
        for index, string in enumerate(self.evaluation_steps, start=1):
            evaluation_steps += f"{index}. {string}\n"
        return evaluation_steps

    def is_successful(self) -> bool:
        if self.error is not None:
            self.success = False
        else:
            try:
                self.success = self.score >= self.threshold
            except TypeError:
                self.success = False
        return self.success

    @property
    def __name__(self):
        if self._include_g_eval_suffix:
            return f"{self.name} [Conversational GEval]"
        else:
            return self.name

File: typescript/src/metrics/conversational-g-eval/conversational-g-eval.ts
import { BaseConversationalMetric } from "@/metrics/base-conversational-metric";
import { resolveThreshold } from "@/metrics/base-metrics";
import { ConversationalTestCase, MultiTurnParams } from "@/test-case";
import { DeepEvalBaseLLM } from "@/models";
import {
  initializeModel,
  generateWithSchema,
  constructVerboseLogs,
  prettifyList,
} from "@/metrics/utils";
import {
  checkConversationalTestCaseParams,
  convertTurnToDict,
} from "@/metrics/conversational-utils";
import { StepsSchema } from "@/metrics/g-eval/schema";
import {
  type Rubric,
  evaluateGEvalPrompt,
  numberEvaluationSteps,
  formatRubrics,
  validateAndSortRubrics,
  validateCriteriaAndEvaluationSteps,
} from "@/metrics/g-eval/utils";
import {
  constructConversationalGEvalTurnParamsString,
  constructNonTurnsTestCaseString,
} from "@/metrics/conversational-g-eval/utils";
import { type MetricTemplateOverride } from "@/templates/override";

const TEMPLATE_CLASS = "ConversationalGEval";

export type ConversationalGEvalTemplateOverride =
  MetricTemplateOverride<"ConversationalGEval">;

export interface ConversationalGEvalMetricOptions {
  name: string;
  /** Which turn/conversation fields the judgement considers (CONTENT + ROLE always added). */
  evaluationParams: MultiTurnParams[];
  criteria?: string;
  evaluationSteps?: string[];
  rubric?: Rubric[];
  model?: DeepEvalBaseLLM | string;
  threshold?: number | null;
  /** Score-token alternatives to weigh, on models that report log probabilities. */
  topLogprobs?: number;
  flaky?: boolean;
  strictMode?: boolean;
  verboseMode?: boolean;
  showIndicator?: boolean;
  includeGEvalSuffix?: boolean;
  evaluationTemplate?: ConversationalGEvalTemplateOverride;
}

/**
 * Conversational GEval — a flexible LLM judge over a whole conversation: generate
 * eval steps from `criteria` (or use supplied `evaluationSteps`), then score the
 * conversation 0–10 (normalized to 0–1) with a reason. Like single-turn GEval but
 * over turns + conversation-level fields.
 */
export class ConversationalGEval extends BaseConversationalMetric {
  evaluationParams: MultiTurnParams[];
  criteria?: string;
  evaluationSteps?: string[];
  rubric?: Rubric[];
  readonly metricName: string;
  private readonly includeGEvalSuffix: boolean;
  private readonly topLogprobs: number;

  constructor(options: ConversationalGEvalMetricOptions) {
    if (!options.evaluationParams || options.evaluationParams.length === 0) {
      throw new Error("evaluationParams cannot be an empty list.");
    }
    if (options.criteria != null || options.evaluationSteps != null) {
      validateCriteriaAndEvaluationSteps(
        options.criteria,
        options.evaluationSteps,
      );
    }
    const strictMode = options.strictMode ?? false;
    super(strictMode ? 1 : resolveThreshold(options.threshold, 0.5), {
      strictMode,
      verboseMode: options.verboseMode,
      showIndicator: options.showIndicator,
      flaky: options.flaky,
      evaluationTemplate: options.evaluationTemplate,
    });
    this.templateClass = TEMPLATE_CLASS;

    // CONTENT + ROLE are always required (mirrors Python).
    const params = [...options.evaluationParams];
    if (!params.includes(MultiTurnParams.CONTENT))
      params.push(MultiTurnParams.CONTENT);
    if (!params.includes(MultiTurnParams.ROLE))
      params.push(MultiTurnParams.ROLE);
    this.evaluationParams = params;
    this.requiredParams = params;

    this.metricName = options.name;
    this.criteria = options.criteria;
    this.rubric = validateAndSortRubrics(options.rubric);
    this.evaluationSteps =
      options.evaluationSteps && options.evaluationSteps.length > 0
        ? options.evaluationSteps
        : undefined;
    this.includeGEvalSuffix = options.includeGEvalSuffix ?? true;
    this.topLogprobs = options.topLogprobs ?? 20;

    const { model, usingNativeModel } = initializeModel(options.model);
    this.model = model;
    this.usingNativeModel = usingNativeModel;
    this.evaluationModel = this.model.getModelName();
  }

  async measure(testCase: ConversationalTestCase): Promise<number> {
    this.error = undefined;
    await this.startProgress();
    try {
      checkConversationalTestCaseParams(testCase, this.requiredParams, this);
      this.evaluationCost = this.usingNativeModel ? 0 : undefined;

      this.evaluationSteps = await this.generateEvaluationSteps();
      const [gScore, reason] = await this.evaluate(testCase);

      this.score = this.applyStrictMode(gScore / 10);
      this.reason = reason;
      this.success = this.isSuccessful();

      this.verboseLogs = constructVerboseLogs(this, [
        `Criteria:\n${this.criteria}`,
        `Evaluation Steps:\n${prettifyList(this.evaluationSteps)}`,
        `Rubric:\n${formatRubrics(this.rubric)}`,
        `Score: ${this.score}\nReason: ${this.reason}`,
      ]);
      return this.score;
    } finally {
      this.stopProgress();
    }
  }

  private async generateEvaluationSteps(): Promise<string[]> {
    if (this.evaluationSteps) return this.evaluationSteps;
    const prompt = this.getPrompt("generate_evaluation_steps", {
      criteria: this.criteria,
      parameters: constructConversationalGEvalTurnParamsString(
        this.evaluationParams,
      ),
    });
    const { steps } = await generateWithSchema(this, prompt, StepsSchema);
    return steps;
  }

  private async evaluate(
    testCase: ConversationalTestCase,
  ): Promise<[number, string]> {
    const prompt = this.getPrompt("generate_evaluation_results", {
      evaluation_steps: numberEvaluationSteps(this.evaluationSteps ?? []),
      test_case_content: constructNonTurnsTestCaseString(
        this.evaluationParams,
        testCase,
      ),
      turns: testCase.turns.map((t) =>
        convertTurnToDict(t, this.evaluationParams),
      ),
      parameters: constructConversationalGEvalTurnParamsString(
        this.evaluationParams,
      ),
      rubric: this.rubric ? formatRubrics(this.rubric) : null,
    });
    return evaluateGEvalPrompt(this, prompt, {
      topLogprobs: this.topLogprobs,
      strictMode: this.strictMode,
    });
  }

  get name(): string {
    return this.includeGEvalSuffix
      ? `${this.metricName} [Conversational GEval]`
      : this.metricName;
  }
}
